plugins/module_utils/mso.py

```diff
@@ -294,96 +327,119 @@ def __init__(self, module):
+        # Set NDO params
         else:
             self.connection = Connection(self.module._socket_path)
             if self.connection.get_platform() == "cisco.nd":
                 self.platform = "nd"
 
+            # Set base_uri
+            self.baseuri = "{0}mso/api/v1/".format(self.base_only_uri)
+
+        if self.params.get("password"):
+            # Perform password-based authentication, log on using password
+            self.login()
+        else:
+            self.fail_json(msg="Parameter 'password' is required for authentication")
+
     def get_login_domain_id(self, domain):
-        ''' Get a domain and return its id '''
+        """Get a domain and return its id"""
         if domain is None:
             return domain
-        d = self.get_obj('auth/login-domains', key='domains', name=domain)
+        d = self.get_obj("auth/login-domains", key="domains", name=domain)
         if not d:
             self.fail_json(msg="Login domain '%s' is not a valid domain name." % domain)
-        if 'id' not in d:
+        if "id" not in d:
             self.fail_json(msg="Login domain lookup failed for domain '%s': %s" % (domain, d))
-        return d['id']
+        return d["id"]
 
     def login(self):
-        ''' Log in to MSO '''
+        """Log in to MSO"""
 
         # Perform login request
-        if (self.params.get('login_domain') is not None) and (self.params.get('login_domain') != 'Local'):
-            domain_id = self.get_login_domain_id(self.params.get('login_domain'))
-            payload = {'username': self.params.get('username', 'admin'), 'password': self.params.get('password'), 'domainId': domain_id}
+        if (self.params.get("login_domain") is not None) and (self.params.get("login_domain") != "Local"):
+            domain_id = self.get_login_domain_id(self.params.get("login_domain"))
+            payload = {
+                "username": self.params.get("username", "admin"),
+                "password": self.params.get("password"),
+                "domainId": domain_id,
+            }
         else:
-            payload = {'username': self.params.get('username', 'admin'), 'password': self.params.get('password')}
-        self.url = urljoin(self.baseuri, 'auth/login')
-        resp, auth = fetch_url(self.module,
-                               self.url,
-                               data=json.dumps(payload),
-                               method='POST',
-                               headers=self.headers,
-                               timeout=self.params.get('timeout'),
-                               use_proxy=self.params.get('use_proxy'))
+            payload = {
+                "username": self.params.get("username", "admin"),
+                "password": self.params.get("password"),
+            }
+
+        self.url = urljoin(self.baseuri, "auth/login")
+
+        if self.platform == "nd":
+            self.url = "{0}login".format(self.base_only_uri)
```

Comment:
This login function shouldn't be used for ND so not sure why we are adding this here.